### PR TITLE
lfsapi: re-authenticate HTTP redirects when needed

### PR DESCRIPTION
**Pull request by ttaylorr**
This pull request teaches package `lfsapi` to re-authenticate HTTP requests when the response indicates a redirection.

When an HTTP request is sent through package `lfsapi`, there are a few paths that we can take down into `do()` and `doWithRedirects()`. The one that we're concerned with involves one of two outcomes:

1. We send an HTTP request that is responded to with a 301 or 302 code. If we're within our retry limit, we construct a new request sent to the new host/path, and copy over the headers from the old request \[[1](https://github.com/git-lfs/git-lfs/blob/v2.4.1/lfsapi/client.go#L360-L368)\]. Notably, we do _not_ copy the `Authorization` header iff the new host differs from the old one, as a security concern. We retry the request.

2. Same setup as (1), except we have exceeded our retry count. We complain and fail \[[2](https://github.com/git-lfs/git-lfs/blob/v2.4.1/lfsapi/client.go#L213-L216)\].

In case (1) this security measure has the opportunity to fail when _the redirected host differs from the requested host_ and _the redirected host requires authentication_. Because we recur early on `doWithRedirects()`, and _not_ `DoWithAuth()`, we don't re-authenticate the request, and instead devolve into a case like https://github.com/git-lfs/git-lfs/issues/3025.

But, recurring through `DoWithAuth()` would throw away our state `via`, allowing authenticated request to flow through different pairs of hosts infinitely. That's no good.

Instead, let's try this:

- 1a3a19c: set up `do()` to take `via []*http.Request`, the list of requests that have lead up to this one (redirects), or `nil` if there are no such requests.
- b1c2b42: keep track of the `remote` associated with a given request, or `""` if no remote matches, or we don't care which remote is associated.
- 5257e58: branch through `doWithAuth` and override the `via` list such that we can re-authenticate requests across different hosts.

One tricky thing in 5257e58 is that we don't actually know whether the original request was designated to be authenticated or not. An additional tricky thing is that we might _never_ authenticate the original request, and instead load from `http.[$url.]extraHeaders`, such that `Authorization` header is present, but Git LFS didn't put it there.

In all of those cases, we reduce the problem to checking whether the original request contained an `Authorization` header.

Closes: https://github.com/git-lfs/git-lfs/issues/3025.

##

/cc @git-lfs/core https://github.com/git-lfs/git-lfs/issues/3025

### CODE REVIEW

File: lfsapi/auth.go [hunk 1]
```diff
@@ -57,11 +67,11 @@ func (c *Client) DoWithAuth(remote string, req *http.Request) (*http.Response, e
 	return res, err
 }
 
-func (c *Client) doWithCreds(req *http.Request, credHelper CredentialHelper, creds Creds, credsURL *url.URL, access Access) (*http.Response, error) {
+func (c *Client) doWithCreds(req *http.Request, credHelper CredentialHelper, creds Creds, credsURL *url.URL, access Access, via []*http.Request) (*http.Response, error) {
```
Comment (ashmckenzie) [inline lfsapi/auth.go]: Monster arg list!

**Comment by ashmckenzie** [comment]
Looks good!  I started to prepare a fix but found myself putting `remote` in a lot of places and bailed.  

Challenging one to test!

**Comment by ttaylorr** [comment]
> Challenging one to test!

Thanks for the reminder. We test package `lfsapi` from Go, mostly as a consequence of not having established ways to run "fake" HTTP servers from the integration tests. I added a test exercising this new behavior here: https://github.com/git-lfs/git-lfs/pull/3028/commits/8715e893e3f240f41cfc057234ae7d8ac16d5b28.

**Comment by ttaylorr** [comment]
Windows failures are unrelated.

**Comment by dosire** [comment]
Thanks for fixing this @ttaylorr

**Pull request merged**